keps/sig-auth/2718-20210511-client-exec-proxy/README.md

```diff
@@ -0,0 +1,837 @@
+this could be less portable if we need to use a non UDS communication mechanism
+for windows, or if we support TCP/IP in the future.
+
+### Caching
+
+The connection from the client to the proxy can be cached in the client based
+on the cluster details, the existing HTTP_PROXY configuration, the exec-proxy
+URL, and the certificates used for mTLS.
+
+### Proxy Shutdown
+
+In the case where the client dies unexpectedly, due to a crash or SIGKILL, the
+proxy should also die.  To handle this, the proxy will attempt to take a file
+lock that the parent holds and quit if it ever successfully takes the lock.
+
+Alterntatively, the proxy could quit after a certain amount of idle time.  For
+example, proxies that are idle for more than 5 minutes will be required to
+quit.  Idle quit is simpler, but the lock approach handles the kubectl use case
+better, so it is preferred.
+
+### API
+
+The communication over stdin and stdout with the exec'd binary will be done
+with using the existing ExecCredentialSpec and ExecCredentialStatus objects.
+
+```go
+// ExecCredentialSpec holds request and runtime specific information provided by
+// the transport.
+type ExecCredentialSpec struct {
+	// Cluster contains information to allow an exec plugin to communicate with the
+	// kubernetes cluster being authenticated to. Note that Cluster is non-nil only
+	// when provideClusterInfo is set to true in the exec provider config (i.e.,
+	// ExecConfig.ProvideClusterInfo).
+	// +optional
+	Cluster *Cluster `json:"cluster,omitempty"`
+
+	// Interactive declares whether stdin has been passed to this exec plugin.
+	Interactive bool `json:"interactive"`
+}
+
+// ExecCredentialStatus holds credentials for the transport to use.
+//
+// Token and ClientKeyData are sensitive fields. This data should only be
+// transmitted in-memory between client and exec plugin process. The Exec plugin
+// itself should at least be protected via file permissions.
+type ExecCredentialStatus struct {
+	// ExpirationTimestamp indicates a time when the provided credentials expire.
+	// +optional
+	ExpirationTimestamp *metav1.Time `json:"expirationTimestamp,omitempty"`
+	// Token is a bearer token used by the client for request authentication.
+	Token string `json:"token,omitempty" datapolicy:"token"`
+	// PEM-encoded client TLS certificates (including intermediates, if any).
+	ClientCertificateData string `json:"clientCertificateData,omitempty"`
+	// PEM-encoded private key for the above certificate.
+	ClientKeyData string `json:"clientKeyData,omitempty" datapolicy:"security-key"`
+
+	// New:
+
+	// ProxyConfig is the config returned by the exec-proxy
+	// +optional
+	ProxyConfig *ProxyConfig `json:"proxyConfig,omitempty"`
+}
+
+
+// ProxyConfig configures the client to proxy requests via the returned address.
+type ProxyConfig struct {
+	// URL is the location of the exec proxy listener.  Supports unix domain
+	// socket locations in the format unix:///path/to/socket.
```

Comment:
Do we allow unnamed sockets? Do we allow anonymous sockets?

Reply:
We probably should stick with simple file based sockets.  I am unsure if the other ones even work on windows.

Reply:
Is there a difference between unnamed and anonymous sockets?  Between unnamed, abstract and named/pathnamed, it looks like all are actually supported by the windows implementation of `AF_UNIX` [1].  Abstract might actually make sense for us because it avoids the use of the filesystem (but they still have a name), and if we are not relying on the filesystem for security then it simplifies the implementation by avoiding path collisions and avoiding the need to clean up the socket.  Unnamed doesn't make sense to me because it would require a different API and the socket would have to be passed over `fork()`.  I think that would be a more difficult API for plugin implementers.

[1]: https://devblogs.microsoft.com/commandline/af_unix-comes-to-windows/

Reply:
Sorry, meant to say abstract, not anonymous.

Reply:
https://github.com/microsoft/WSL/issues/4240

Reply:
https://github.com/grpc/grpc/issues/22285#issuecomment-1169211526
https://github.com/microsoft/WSL/issues/5272#issuecomment-1236620129

Reply:
Good find @enj

Reply:
> [microsoft/WSL#4240 (comment)](https://github.com/microsoft/WSL/issues/4240#issuecomment-620805115)

I have confirmed with `@sunilmut` that this is still true.